ui/src/components/build/CreateBuildConfigDialog.js

```diff
@@ -0,0 +1,189 @@
+      KEY_CR_SSH_AUTH_NAME,
+      KEY_CR_SSH_PRIVATE_KEY
+    ]);
+    this.sourceState = new SubState(this, 'config.source', sourceValidation, [KEY_CR_SOURCE_GITURL]);
+    this.androidCredentialsState = new SubState(this, 'config.build.androidCredentials', androidCredentialsValidation, [
+      KEY_CR_BUILD_ANDROID_CREDENTIALS_NAME,
+      KEY_CR_BUILD_ANDROID_CREDENTIALS_KEYSTORE,
+      KEY_CR_BUILD_ANDROID_CREDENTIALS_KEYSTORE_ALIAS,
+      KEY_CR_BUILD_ANDROID_CREDENTIALS_KEYSTORE_PASSWORD
+    ]);
+    this.iOSCredentialsState = new SubState(this, 'config.build.iosCredentials', iOSCredentialsValidation, [
+      KEY_CR_BUILD_IOS_CREDENTIALS,
+      KEY_CR_BUILD_IOS_CREDENTIALS_DEVELOPER_PROFILE,
+      KEY_CR_BUILD_IOS_CREDENTIALS_NAME
+    ]);
+  }
+
+  componentDidUpdate(prevProps) {
+    this.validate();
+    const { initialConfig: prevConfig = {} } = prevProps;
+    const { initialConfig = {} } = this.props;
+    if (!isEqual(prevConfig, initialConfig)) {
+      this.configState.set(KEY_CR_CLIENT_ID, initialConfig[KEY_CR_CLIENT_ID]);
+      this.configState.set(KEY_CR_CLIENT_TYPE, initialConfig[KEY_CR_CLIENT_TYPE]);
+      switch (initialConfig[KEY_CR_CLIENT_TYPE]) {
+        case CLIENT_TYPE_ANDROID:
+          this.buildState.set(KEY_CR_BUILD_PLATFORM, BUILD_PLATFORM_ANDROID);
+          this.setState({ [KEY_HIDE_PLATDORM]: true });
+          break;
+        case CLIENT_TYPE_IOS:
+          this.buildState.set(KEY_CR_BUILD_PLATFORM, BUILD_PLATFORM_IOS);
+          this.setState({ [KEY_HIDE_PLATDORM]: true });
+          break;
+        default:
+          this.setState({ [KEY_HIDE_PLATDORM]: false });
+      }
+    }
+  }
+
+  validate() {
+    let validVal = this.configState.isAllValid() && this.buildState.isAllValid() && this.sourceState.isAllValid();
+    if (this.sourceState.get(KEY_CR_SOURCE_AUTH_TYPE) === BUILD_AUTH_TYPE_BASIC)
+      validVal = validVal && this.basicAuthState.isAllValid();
+    if (this.sourceState.get(KEY_CR_SOURCE_AUTH_TYPE) === BUILD_AUTH_TYPE_SSH)
+      validVal = validVal && this.sshAuthState.isAllValid();
+    if (this.state.valid !== validVal) this.setState({ valid: validVal });
+  }
+
+  onSaveBuildConfig = () => {
+    const { config } = this.state;
+    const { onSave, update = false, initialConfig } = this.props;
+    onSave && onSave({ update, config: { ...initialConfig, ...config } });
+  };
+
+  cancel = () => {
+    const { onCancel } = this.props;
+    onCancel && onCancel();
+  };
+
+  renderFormButtons = (onCancel, update, valid) => (
+    <div>
+      <Button bsStyle="primary" onClick={this.onSaveBuildConfig} disabled={!valid}>
+        {update ? 'Save' : 'Create'}
+      </Button>
+      <Button onClick={onCancel}>Cancel</Button>
+    </div>
+  );
+
+  getValidationState = () => {};
+
+  render() {
+    const { show, title, onCancel, update } = this.props;
+    const { valid = false } = this.state;
+    return (
+      <Modal show={show} container={this}>
```

Comment:
If we use this `container` property should we also add these [1] css styles?

[1] https://react-bootstrap.github.io/components/modal/#modals-contained

Reply:
ok